config/config.go

```diff
@@ -1,18 +1,275 @@
+	if d.Template.absDirectory != "" {
+		return d.Template.absDirectory
+	}
+	switch true {
+	case filepath.IsAbs(d.TemplatesDir):
+		d.Template.absDirectory = filepath.Join(d.TemplatesDir, d.Template.Name)
+		return d.Template.absDirectory
+	case filepath.IsAbs(ConfigFilePath):
+		d.Template.absDirectory = filepath.Join(filepath.Dir(ConfigFilePath), d.TemplatesDir, d.Template.Name)
+		return d.Template.absDirectory
+
+	// No absolute dir provided. Let's build one
+	default:
+		d.Template.absDirectory = filepath.Join(ProcWorkingDir, filepath.Dir(ConfigFilePath), d.TemplatesDir, d.Template.Name)
+		return d.Template.absDirectory
+	}
 }
 
-type GeneratorConfig struct {
-	ModuleName        string
-	BasePackageName   string
-	TemplateDirectory string
-	TemplateName      string
+// GetOutputDir is used by the render engine to determine where rendered files will be written to
+func (d *Data) GetOutputDir() (path string) {
+	if d.absOutputPath != "" {
+		return d.absOutputPath
+	}
+	switch true {
+	case filepath.IsAbs(d.OutputPath):
+		d.absOutputPath = d.OutputPath
+		return d.absOutputPath
+	case filepath.IsAbs(ConfigFilePath):
+		d.absOutputPath = filepath.Join(
+			filepath.Dir(ConfigFilePath),
+			d.OutputPath,
+		)
+		return d.absOutputPath
+
+	// No absolute dir provided. Let's build one
+	default:
+		d.absOutputPath = filepath.Join(
+			ProcWorkingDir,
+			filepath.Dir(ConfigFilePath),
+			d.OutputPath,
+		)
+		return d.absOutputPath
+	}
+}
+
+// GetOpenAPIFilepath uses the current working directory, resolved config file and the openAPI file that was specified
+// in the config file to determine the absolute path to an OpenAPI file. It takes into account that any of these,
+// except the working directory could be relative. It returns the absolute value on every call by caching the result of
+// it's first run and returning that on successive calls
+func (d *Data) GetOpenAPIFilepath() (path string) {
+	if d.absOpenAPIPath != "" { // we shouldn't have to run below logic multiple times
+		return d.absOpenAPIPath
+	}
+	if filepath.IsAbs(d.OpenAPIFile) {
+		d.absOpenAPIPath = d.OpenAPIFile
+		return d.absOpenAPIPath
+	} else {
+		if filepath.IsAbs(ConfigFilePath) {
+			d.absOpenAPIPath = filepath.Join(
+				filepath.Dir(ConfigFilePath),
+				d.OpenAPIFile,
+			)
+			return d.absOpenAPIPath
+		} else {
+			d.absOpenAPIPath = filepath.Join(
+				// TODO test this
+				// not relative as per above comment
+				ProcWorkingDir,
+				filepath.Dir(ConfigFilePath),
+				d.OpenAPIFile,
+			)
+			return d.absOpenAPIPath
+		}
+	}
+}
+
+func (d *Data) validTemplateName() bool {
+	dirs, err := d.validTemplates()
+	if err != nil {
+		fmt.Println(fmt.Errorf("could not list valid templates in %s :: %w", d.TemplatesDir, err))
+		return false
+	}
+	for _, dir := range dirs {
+		if dir == d.Template.Name {
+			return true
+		}
+	}
+	return false
+}
+
+func (d *Data) validTemplates() (templates []string, err error) {
+	dirs, err := os.ReadDir(d.TemplatesDir)
```

Comment:
you're relying on state here that may not always be set. This can cause hard to troubleshoot errors that require following potentially complex codepaths and tracking multiple function calls to debug.

Is there a way to guarantee that these paths have been built before calling these functions?

Reply:
The program simply crashes as a result of trying to list these directories and if an error is bubbled up in any of the cases that this is attempted, we crash with a descriptive error message. So I'm not convinced that the errors will be in any way vague to the user.